src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+            /// </remarks>
+            NotRecognized,
+
+            /// <summary>
+            /// A request made by a caller that is already in the semaphore is immediately executed,
+            /// and shares the same semaphore slot with its parent.
+            /// This nested request must exit before its parent (Strict LIFO/stack behavior).
+            /// Exiting the semaphore before a child has or after the parent has will cause an
+            /// <see cref="InvalidOperationException"/> to fault the <see cref="Task"/> returned
+            /// from <see cref="ExecuteAsync(Func{Task}, CancellationToken)"/>.
+            /// </summary>
+            /// <remarks>
+            /// When reentrancy is a requirement, this mode helps ensure that reentrancy only happens
+            /// where code enters a semaphore, then awaits on other code that itself may enter the semaphore.
+            /// When a violation occurs, this semaphore transitions into a faulted state, after which any call
+            /// will throw an <see cref="InvalidOperationException"/>.
+            /// </remarks>
+            Stack,
+
+            /// <summary>
+            /// A request made by a caller that is already in the semaphore is immediately executed,
+            /// and shares the same semaphore slot with its parent.
+            /// The slot is only released when all requests have exited, which may be in any order.
+            /// </summary>
+            /// <remarks>
+            /// This is the most permissive, but has the highest risk that leaked semaphore access will remain undetected.
+            /// Leaked semaphore access is a condition where code is inappropriately considered parented to another semaphore holder,
+            /// leading to it being allowed to run code within the semaphore, potentially in parallel with the actual semaphore holder.
+            /// </remarks>
+            Freeform,
+        }
+
+        /// <summary>
+        /// Gets the number of openings that remain in the semaphore.
+        /// </summary>
+        public int CurrentCount
+        {
+            get
+            {
+                this.ThrowIfFaulted();
+                return this.semaphore.CurrentCount;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ReentrantSemaphore"/> class.
+        /// </summary>
+        /// <param name="initialCount">The initial number of concurrent operations to allow.</param>
+        /// <param name="joinableTaskContext">The <see cref="JoinableTaskContext"/> to use to mitigate deadlocks.</param>
+        /// <param name="mode">How to respond to a semaphore request by a caller that has already entered the semaphore.</param>
+        public static ReentrantSemaphore Create(int initialCount = 1, JoinableTaskContext joinableTaskContext = default, ReentrancyMode mode = ReentrancyMode.NotAllowed)
```

Comment:
>mode [](start = 136, length = 4)

Does it make sense to have a default mode?  It might be beneficial to force the caller to be more intentional about the choice of behavior.

Reply:
Great question. If we made that a mandatory parameter, we'd need to make JTC mandatory as well or reorder the parameters to put JTC last. I'm divided on this point... still mulling it over.